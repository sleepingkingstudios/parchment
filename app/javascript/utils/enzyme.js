import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
