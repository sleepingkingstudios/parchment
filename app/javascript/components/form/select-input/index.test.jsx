import React from 'react';
import { shallow } from 'enzyme';

import FormSelectInput from './index';

describe('<FormSelectInput />', () => {
  const onChange = () => {};
  const options = [
    { label: 'Option 1', value: 'one' },
    { label: 'Option 2', value: 'two' },
    { label: 'Option 3', value: 'three' },
  ];
  const props = {
    id: 'property-name',
    options,
    prop: 'propertyName',
    value: 'Property Value',
    onChange,
  };
  const rendered = shallow(<FormSelectInput {...props} />);

  it('should create the select input', () => {
    const { id, prop, value } = props;

    expect(rendered).toHaveDisplayName('select');
    expect(rendered).toHaveClassName('custom-select');
    expect(rendered).toHaveProp('id', id);
    expect(rendered).toHaveProp('value', value);
    expect(rendered).toHaveProp('onChange', onChange);
    expect(rendered).toHaveProp('data-prop-name', prop);
  });

  it('should create the select options', () => {
    const optionElements = rendered.find('FormSelectOption');

    expect(optionElements.length).toEqual(options.length);

    optionElements.forEach((element, index) => {
      const { label, value } = options[index];

      expect(element).toHaveProp('label', label);
      expect(element).toHaveProp('value', value);
    });
  });

  describe('with defaultOption: value', () => {
    const defaultOption = 'Default Option';
    const propsWithDefault = {
      id: 'property-name',
      defaultOption,
      options,
      prop: 'propertyName',
      value: 'Property Value',
      onChange,
    };
    const renderedWithDefault = shallow(<FormSelectInput {...propsWithDefault} />);

    it('should create the select options', () => {
      const optionElements = renderedWithDefault.find('FormSelectOption');
      const expectedOptions = options.map(item => item);

      expectedOptions.unshift({ label: defaultOption, value: '' });

      expect(optionElements.length).toEqual(expectedOptions.length);

      optionElements.forEach((element, index) => {
        const { label, value } = expectedOptions[index];

        expect(element).toHaveProp('label', label);
        expect(element).toHaveProp('value', value);
      });
    });
  });
});
