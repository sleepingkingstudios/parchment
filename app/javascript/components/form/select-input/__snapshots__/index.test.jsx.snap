// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSelectInput /> should match the snapshot 1`] = `
<select
  className="custom-select"
  id="property-name"
  onChange={[Function]}
  value="Property Value"
>
  <FormSelectOption
    key="option-0"
    label="Option 1"
    value="one"
  />
  <FormSelectOption
    key="option-1"
    label="Option 2"
    value="two"
  />
  <FormSelectOption
    key="option-2"
    label="Option 3"
    value="three"
  />
</select>
`;
